iree/turbine/kernel/wave/hoisting.py

```diff
@@ -13,25 +15,85 @@
 logger = get_logger("turbine.wave.hoisting")
 
 
-def get_allocs(graph: fx.Graph) -> list[CustomOp]:
-    return [
-        custom_node
-        for node in graph.nodes
-        if isinstance((custom_node := get_custom(node)), Allocate)
-    ]
+def get_hoistable_ops(
+    graph: fx.Graph,
+    captured_vars: list[CustomOp],
+    induction_variable: IndexExpr,
+) -> list[CustomOp]:
+    """
+    Get hoistable ops. Currently only handle allocs and read who doesn't depends on
+    induction variables.
+
+    Note: For codegen to work properly, we'd need to hoist allocs first. This is to avoid
+    using alloc before defined/non-dominating behavior.
+    (e.g hoisting read from global to shared before shared alloc is defined.)
+    """
+    hoistable_allocs = []
+    hoistable_ops = []
+    for node in graph.nodes:
+        custom_node = get_custom(node)
+        if isinstance(custom_node, Allocate):
+            hoistable_allocs.append(custom_node)
+        elif isinstance(custom_node, Read):
+            if custom_node.index is None:
+                continue
+            # Only handle case where memory is captured var.
+            # i.e it has source value from root graph.
+            if not custom_node.memory in captured_vars:
+                continue
+            # Only handle case where we are not writing to the same memory.
+            # Counterproof: we may expect different read if we write to same memory.
+            for memory_user in custom_node.memory.users:
+                if isinstance(get_custom(memory_user), Write):
+                    continue
+            # Only hoist Read that is loop invariant.
+            dims_indexing = [ind.start for ind in custom_node.index.values()]
+            dim_depends_on_ivar = [ind.has(induction_variable) for ind in dims_indexing]
+            if any(dim_depends_on_ivar):
+                continue
+            hoistable_ops.append(custom_node)
+        else:
+            continue
+    all_hoistables_ops = hoistable_allocs + hoistable_ops
```

Comment:
I think we can unify them into single list

Reply:
The main reason why it is important to have two list is because we want to ensure that allocs are processed/hoisted out first. This is to avoid using alloc before defined/non-dominating behavior. (e.g hoisting read from global to shared before shared alloc is defined.)

Reply:
oh I missed the other comment, yes in that case makes sense to separate them out.